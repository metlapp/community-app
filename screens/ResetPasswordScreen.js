import React, { useState} from "react";
import axios from "axios";
import {SafeAreaView, View} from "react-native";
import {Title} from "react-native-paper";
import * as Yup from "yup";

import AppFormField from "../components/AppFormField";
import defaultStyles from "../config/defaultStyles";
import ErrorMessage from "../components/ErrorMessage";
import Form from "../components/Form";
import SubmitButton from "../components/SubmitButton";

import {apiConfig} from "../config/config";

const validationSchema = Yup.object().shape({
  password: Yup.string().trim().required().min(8).label("New Password"),
  confirmPassword: Yup.string().trim().oneOf([Yup.ref('password'), null]).required().min(8).label("Confirm Password"),
});

export default function ResetPasswordScreen({ navigation, route }) {
  const [errorVisible, setErrorVisible] = useState(false);
  const [error, setError] = useState("");

  // Getting token from emailed URL
  const { token } = route.params;

  const savePassword = async (values) => {
    setErrorVisible(false);

    await axios
      .post(apiConfig.baseUrl + 'api/password_reset/confirm/', {token, password: values.password})
      .then((data) => {
        // if password reset was successful, go to login screen
        if (data.status === 200) {
          navigation.navigate("Login");
        }
      })
      .catch((err) => {
        switch (err.response.status) {
          // setting error based on status code
          case 404:
            setErrorVisible(true);
            setError(
              "Invalid token. Please request a new password reset email."
            );
            return;
          case 400:
            setErrorVisible(true);
            setError(
              "Password cannot be similar to your first name or be a previously used password. Please try again."
            );
            return;
          default:
            navigation.navigate("Login");
        }
        setErrorVisible(true);
        setError(
          "Can not connect to server, ensure you have a working internet connection and try again"
        );
      });
  };

  return (
    <SafeAreaView>
      <View style={defaultStyles.mainContainer}>

        <View style={defaultStyles.formContainer}>

          <Form
            initialValues={{password: "", confirmPassword: ""}}
            onSubmit={(values) => {
              savePassword(values);
            }}
            validationSchema={validationSchema}
          >
            <Title style={defaultStyles.formFieldTitle}>Password</Title>
            <AppFormField
              name="password"
              placeholder="New password"
              secureTextEntry
              testID="pass"
            />
            <Title style={defaultStyles.formFieldTitle}>Confirm Password</Title>
            <AppFormField
              name="confirmPassword"
              placeholder="Repeat new password"
              secureTextEntry
              testID="confirmPass"
            />
            <ErrorMessage
              error={error}
              styling={{justifyContent: "center", height: 60, fontSize: 20}}
              visible={errorVisible}
            />
            <SubmitButton title="Save"/>
          </Form>
        </View>
      </View>
    </SafeAreaView>
  );
}
