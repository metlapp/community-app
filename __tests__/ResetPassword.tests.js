import "react-native";
import React from "react";
import ResetPasswordScreen from "../screens/ResetPasswordScreen";
import { render, fireEvent, act, waitFor } from "@testing-library/react-native";
import AuthContext from "../auth/Context";

describe("<ResetPasswordScreen />", () => {
  it("User can change their Email with a valid Email", async () => {
    const setUser = jest.fn();
    const user = { email: "email@email.com" };

    const { getByText, getByTestId } = render(
      <AuthContext.Provider value={{ user, setUser }}>
        <ResetPasswordScreen />
      </AuthContext.Provider>
    );

    act(() => {
      fireEvent(getByTestId("pass"), "onChangeText", "password12");
    });
    act(() => {
      fireEvent(getByTestId("confirmPass"), "onChangeText", "password12");
    });
    act(() => {
      fireEvent(getByText("Save"), "onPress");
    });
    await waitFor(() => {
      expect(setUser).toHaveBeenCalled();
    });
  });

  it("User can change their Email with a valid Email", async () => {
    const setUser = jest.fn();
    const user = { email: "email@email.com" };

    const { getByText, getByTestId } = render(
      <AuthContext.Provider value={{ user, setUser }}>
        <ResetPasswordScreen />
      </AuthContext.Provider>
    );

    act(() => {
      fireEvent(getByTestId("pass"), "onChangeText", "password13");
    });
    act(() => {
      fireEvent(getByTestId("confirmPass"), "onChangeText", "password12");
    });
    act(() => {
      fireEvent(getByText("Save"), "onPress");
    });
    await waitFor(() => {
      expect(setUser).not.toHaveBeenCalled();
    });
  });
});
