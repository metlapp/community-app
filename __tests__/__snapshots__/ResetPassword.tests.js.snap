// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetPasswordScreen /> matches the snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": "100%",
        "marginHorizontal": 40,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "marginTop": 40,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "500",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 20,
                  "letterSpacing": 0.15,
                  "lineHeight": 30,
                  "marginVertical": 2,
                },
                Object {
                  "color": "#f89232",
                  "fontSize": 14,
                  "fontWeight": "bold",
                },
              ],
              Array [
                Object {
                  "fontSize": 20,
                  "letterSpacing": 0.15,
                  "lineHeight": 30,
                  "marginVertical": 2,
                },
                Object {
                  "color": "#f89232",
                  "fontSize": 14,
                  "fontWeight": "bold",
                },
              ],
            ],
          ]
        }
      >
        Password
      </Text>
      <View
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgb(231, 231, 231)",
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
              },
              Object {
                "backgroundColor": "#fff",
                "marginBottom": 20,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgba(0, 0, 0, 0.26)",
                "bottom": 0,
                "height": 2,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "transform": Array [
                  Object {
                    "scaleY": 0.5,
                  },
                ],
              }
            }
          />
          <View
            style={
              Object {
                "minHeight": 45,
                "paddingBottom": 0,
                "paddingTop": 0,
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              editable={true}
              multiline={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="New password"
              placeholderTextColor="#ccc"
              rejectResponderTermination={true}
              secureTextEntry={true}
              selectionColor="#6200ee"
              style={
                Array [
                  Object {
                    "flexGrow": 1,
                    "margin": 0,
                    "textAlign": "left",
                    "zIndex": 1,
                  },
                  Object {
                    "paddingLeft": 0,
                    "paddingRight": 0,
                  },
                  Object {
                    "height": 45,
                  },
                  Object {
                    "paddingBottom": 0,
                    "paddingTop": 0,
                  },
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "textAlignVertical": "center",
                  },
                  Array [
                    Object {},
                  ],
                ]
              }
              testID="pass"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "500",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 20,
                  "letterSpacing": 0.15,
                  "lineHeight": 30,
                  "marginVertical": 2,
                },
                Object {
                  "color": "#f89232",
                  "fontSize": 14,
                  "fontWeight": "bold",
                },
              ],
              Array [
                Object {
                  "fontSize": 20,
                  "letterSpacing": 0.15,
                  "lineHeight": 30,
                  "marginVertical": 2,
                },
                Object {
                  "color": "#f89232",
                  "fontSize": 14,
                  "fontWeight": "bold",
                },
              ],
            ],
          ]
        }
      >
        Confirm Password
      </Text>
      <View
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgb(231, 231, 231)",
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
              },
              Object {
                "backgroundColor": "#fff",
                "marginBottom": 20,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgba(0, 0, 0, 0.26)",
                "bottom": 0,
                "height": 2,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "transform": Array [
                  Object {
                    "scaleY": 0.5,
                  },
                ],
              }
            }
          />
          <View
            style={
              Object {
                "minHeight": 45,
                "paddingBottom": 0,
                "paddingTop": 0,
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              editable={true}
              multiline={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="Repeat new password"
              placeholderTextColor="#ccc"
              rejectResponderTermination={true}
              secureTextEntry={true}
              selectionColor="#6200ee"
              style={
                Array [
                  Object {
                    "flexGrow": 1,
                    "margin": 0,
                    "textAlign": "left",
                    "zIndex": 1,
                  },
                  Object {
                    "paddingLeft": 0,
                    "paddingRight": 0,
                  },
                  Object {
                    "height": 45,
                  },
                  Object {
                    "paddingBottom": 0,
                    "paddingTop": 0,
                  },
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "textAlignVertical": "center",
                  },
                  Array [
                    Object {},
                  ],
                ]
              }
              testID="confirmPass"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.12)",
            "borderColor": "transparent",
            "borderRadius": 6,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 0,
            "marginTop": 20,
            "minWidth": 64,
            "padding": 10,
            "width": "100%",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 6,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    undefined,
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.32)",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              <Text
                style={
                  Object {
                    "color": "#fff",
                    "fontSize": 16,
                    "textTransform": "none",
                  }
                }
              >
                Save
              </Text>
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
